lib/client.ml

```diff
@@ -95,6 +130,171 @@ let route client router msg =
+     and ask him/them to ping B.*)
+  (* This function return the random peer, to which we will ask to ping the first peer *)
+  let rec pick_random_neighbors neighbors number_of_neighbors =
+    let addresses = neighbors |> Base.Hashtbl.keys |> knuth_shuffle in
+    match addresses with
+    | [] -> failwith "pick_random_peers"
+    | elem :: _ ->
+      if number_of_neighbors = 1 then
+        [elem]
+      else
+        elem :: pick_random_neighbors neighbors (number_of_neighbors - 1)
+
+  let update_neighbor_status peer neighbor status =
+    let open Peer in
+    let local_neighbor = get_neighbor peer neighbor.address in
+    match local_neighbor with
+    | Some neighbor ->
+      neighbor.status <- status;
+      Base.Hashtbl.update peer.neighbors neighbor.address ~f:(fun _ -> neighbor)
+    | None -> ()
+  let create_message client message (recipient : Peer.t) =
+    Message.
+      {
+        category = Failure_detection;
+        id = -1;
+        sender = address_of !client;
+        recipient = recipient.address;
+        payload = Util.Encoding.pack bin_writer_message message;
+      }
+
+  let send_message message client (recipient : Peer.t) =
+    let message = create_message client message recipient in
+    send_to client message
+
+  let send_ping_to client peer = send_message Ping client peer
+
+  let send_acknowledge_to client peer = send_message Acknowledge client peer
+
+  let send_ping_request_to client (recipient : Peer.t) =
+    send_message (PingRequest recipient.address) client recipient
+
+  (** Processes an incoming message bound for the failure detector of a client *)
+  let handle_message client message =
+    let open Message in
+    let peer_of_client = peer_from !client in
+    let peer = Peer.from message.sender in
+    let msg = Util.Encoding.unpack bin_read_message message.payload in
+    let t = !client.failure_detector in
+    match msg with
+    | Ping -> send_acknowledge_to client peer
+    | PingRequest addr -> (
+      let new_seq_no = next_seq_no t in
+      let peer_opt = Peer.get_neighbor peer_of_client addr in
+      match peer_opt with
+      | Some peer -> send_ping_request_to client peer
+      | None -> (
+        match%lwt wait_ack_timeout t new_seq_no t.config.protocol_period with
+        | Ok _ -> Lwt.return ()
+        | Error _ -> send_acknowledge_to client peer))
+    | Acknowledge ->
+    match Base.Hashtbl.find t.acknowledges t.sequence_number with
+    | Some cond ->
+      Lwt_condition.broadcast cond ();
+      Lwt.return ()
+    | None -> Lwt.return ()
+
+  (** This function will be called by failure_detection 
+  at each round of the protocol, and update the peers *)
+  let probe_peer t client peer_to_update =
+    let peer_of_client = peer_from !client in
+    let new_seq_no = next_seq_no t in
+    let _ = send_ping_to client peer_to_update in
+    match%lwt wait_ack_timeout t new_seq_no t.config.round_trip_time with
+    | Ok _ -> Lwt.return ()
```

Comment:
Are we missing an action here?
I think we should update the status of the Ack sender (recipient of Ping message) to `Alive` if it replies. Maybe this peer was `Suspicious` or `Faulty` before, but since it now replies, it is not the case anymore.

Reply:
Right